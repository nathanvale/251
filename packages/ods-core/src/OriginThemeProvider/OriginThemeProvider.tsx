import React, { createContext, useContext } from "react";
import {
  ThemeProvider as MUIThemeProvider,
  Theme as MuiTheme,
} from "@material-ui/core";

import {
  LinkComponentProps,
  LinkComponentType,
} from "@origin-digital/ods-types";
import { coreMuiTheme } from "@origin-digital/ods-themes";
import { LayoutThemeProvider } from "../_private/components/LayoutThemeProvider/LayoutThemeProvider";

const DefaultLinkComponent = React.forwardRef<
  HTMLAnchorElement,
  LinkComponentProps
>((props, ref) => {
  const { children, ...restProps } = props;
  return (
    <a {...restProps} ref={ref}>
      {children}
    </a>
  );
});
const LinkComponentContext = createContext<LinkComponentType>(
  DefaultLinkComponent
);
export const useLinkComponent = () => useContext(LinkComponentContext);

export interface OriginThemeProviderProps {
  children: React.ReactNode;
  linkComponent?: LinkComponentType;
  muiTheme?: MuiTheme;
}

export const OriginThemeProvider = ({
  children,
  muiTheme = coreMuiTheme,
  linkComponent,
}: OriginThemeProviderProps) => {
  const linkComponentFromContext = useLinkComponent();
  return (
    <MUIThemeProvider theme={muiTheme}>
      <LayoutThemeProvider muiTheme={muiTheme}>
        <LinkComponentContext.Provider
          value={linkComponent || linkComponentFromContext}
        >
          {children}
        </LinkComponentContext.Provider>
      </LayoutThemeProvider>
    </MUIThemeProvider>
  );
};
