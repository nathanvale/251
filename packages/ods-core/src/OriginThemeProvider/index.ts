export { OriginThemeProvider } from "./OriginThemeProvider";
export type OriginThemeProviderProps = import("./OriginThemeProvider").OriginThemeProviderProps;
