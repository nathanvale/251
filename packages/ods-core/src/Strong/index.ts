export { Strong } from "./Strong";
export type StrongProps = import("./Strong").StrongProps;
