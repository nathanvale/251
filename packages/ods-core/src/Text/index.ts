export { Text } from "./Text";
export { TextContextProvider, TextContext } from "./TextContextProvider";
export { useTextStyles, useTextStylesFromContext } from "./useTextStyles";

export type TextProps = import("./Text").TextProps;
export type UseTextProps = import("./TextContextProvider").UseTextProps;
export type TextContextProviderProps = import("./TextContextProvider").TextContextProviderProps;
