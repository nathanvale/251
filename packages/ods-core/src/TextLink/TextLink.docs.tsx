/* eslint-disable react/display-name */
import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Text, TextProps, TextLink } from "..";

export const docs: ComponentDocs<TextProps> = {
  category: "Interaction",
  componentName: "TextLink",
  description: `By default, this component renders a native "a" element. If youâ€™d like to customise the implementation of this component (e.g. to use a React Router link) you should use TextLinkRenderer instead.`,
  propDescriptions: {},
  migrationGuide: false,
  variant: "list",
  examples: [
    {
      label: "Text Link",
      Code: () => (
        <Text>
          <TextLink href="">Text link</TextLink>
        </Text>
      ),
    },
    {
      label: "Text Link in a sentence",
      Code: () => (
        <Text>
          The last word of a sentence is a{" "}
          <TextLink href="">text link</TextLink>.
        </Text>
      ),
    },
    {
      label: "Subtitle Text Link",
      Code: () => (
        <Text variant="subtitle">
          The last word of a sentence is a{" "}
          <TextLink href="">text link</TextLink>.
        </Text>
      ),
    },
    {
      label: "Subtitle Small Text Link",
      Code: () => (
        <Text variant="subtitle-small">
          The last word of a sentence is a{"  "}
          <TextLink href="">text link</TextLink>.
        </Text>
      ),
    },
    {
      label: "Body Small Text Link",
      Code: () => (
        <Text variant="body-small">
          The last word of a sentence is a{" "}
          <TextLink href="">text link</TextLink>.
        </Text>
      ),
    },
  ],
  snippets: [
    {
      label: "Inline link",
      Code: () => (
        <Text>
          <TextLink href="">Link text</TextLink>.
        </Text>
      ),
    },
  ],
};
