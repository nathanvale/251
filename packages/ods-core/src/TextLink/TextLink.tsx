/* eslint-disable jsx-a11y/anchor-has-content */
import React from "react";
import {
  OptionalTrackableProps,
  LinkComponentProps,
} from "@origin-digital/ods-types";
import {
  TextLinkRenderer,
  TextLinkRendererProps,
} from "../TextLinkRenderer/TextLinkRenderer";

export interface TextLinkProps
  extends Omit<TextLinkRendererProps, "children">,
    OptionalTrackableProps,
    Omit<LinkComponentProps, "className" | "style"> {}

export const TextLink = ({ "data-id": dataId, ...props }: TextLinkProps) => (
  <TextLinkRenderer data-id={dataId}>
    {(styleProps) => <a {...props} {...styleProps} />}
  </TextLinkRenderer>
);

TextLink.defaultProps = {
  "data-id": "text-link",
};

TextLink.displayName = "TextLink";
