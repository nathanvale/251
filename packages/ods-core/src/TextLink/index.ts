export { TextLink } from "./TextLink";
export type TextLinkProps = import("./TextLink").TextLinkProps;
