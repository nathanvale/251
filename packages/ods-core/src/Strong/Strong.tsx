import React, { ReactNode, useContext } from "react";
import { useWeightStyles } from "../_private/hooks/typography";
import { TextContext } from "../Text/TextContext";

export interface StrongProps {
  children: ReactNode;
  "data-id"?: string;
}

export const Strong = ({ children, "data-id": dataId }: StrongProps) => {
  const inText = useContext(TextContext);
  return (
    <strong
      data-id={dataId}
      className={useWeightStyles(inText ? "medium" : "bold")}
    >
      {children}
    </strong>
  );
};

Strong.defaultProps = {
  "data-id": "strong",
};

Strong.displayName = "Strong";
