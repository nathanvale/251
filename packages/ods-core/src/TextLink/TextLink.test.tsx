import React from "react";
import { render, fireEvent } from "@origin-digital/ods-testing-library";
import {
  TrackingEventHandler,
  LinkComponentType,
} from "@origin-digital/ods-types";
import { OriginThemeProvider } from "../OriginThemeProvider/OriginThemeProvider";
import { Text } from "../Text/Text";
import { TrackingProvider } from "../TrackingProvider/TrackingProvider";
import { Strong } from "../Strong/Strong";
import { TextLink } from "./TextLink";

test("It can render", () => {
  const { container } = render(
    <Text>
      <TextLink href="someUrl">Text link</TextLink>
    </Text>
  );
  expect(container.firstChild).toMatchSnapshot();
});

it("should call tracking handler when passed into the TrackingProvider", () => {
  const label = (
    <>
      Hello this is <Strong>strong</Strong>
    </>
  );

  const onTrackingCapture: TrackingEventHandler = jest.fn();

  const { getByTestId } = render(
    <TrackingProvider onTrackingCapture={onTrackingCapture}>
      <Text>
        <TextLink href="">{label}</TextLink>
      </Text>
    </TrackingProvider>
  );
  const node = getByTestId("text-link");
  fireEvent.click(node);
  expect(onTrackingCapture).toHaveBeenCalledWith({
    "data-id": "text-link",
    type: "TextLink",
    label: "Hello this is strong",
    postClickState: undefined,
  });
});

test("it handles a custom link implementation", () => {
  const onClick = jest.fn();

  const CustomLink: LinkComponentType = ({ href, ...restProps }) => (
    // eslint-disable-next-line jsx-a11y/anchor-has-content
    <a href={href} onClick={onClick} {...restProps} />
  );

  const { getByText } = render(
    <OriginThemeProvider linkComponent={CustomLink}>
      <Text>
        <TextLink href="https://www.test.com/">Click me</TextLink>
      </Text>
    </OriginThemeProvider>
  );

  const link = getByText("Click me");
  fireEvent.click(link);

  expect(onClick).toBeCalledTimes(1);

  expect(getByText("Click me").closest("a")).toHaveAttribute(
    "href",
    "https://www.test.com/"
  );
});
