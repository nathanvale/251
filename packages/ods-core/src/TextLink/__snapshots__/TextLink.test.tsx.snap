// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It can render 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  max-width: 100%;
  text-align: left;
}

<span
  class="typography-body-variant-basekick-1 typography-neutral-color-2 typography-regular-weight-3 typography-strong-support-4 -span c0"
  data-id="text"
>
  <a
    class="typography-linkReset-6 typography-neutral-tone-link-7"
    data-id="text-link"
    href="someUrl"
  >
    Text link
  </a>
</span>
`;
